### ISSUE TITLE
EMA C++: Destructor of OmmConsumer occasionally hangs forever in deadlock

### ISSUE THREAD

**Issue by miczuc** [Refinitiv/Real-Time-SDK#138]
Hello.
Destruction of a OmmConsumer object sometimes hangs forever. I examined this situation on a Linux system with pstack ( = gstack). The explanation for the endless stop is as follows:
The destructor of OmmConsumer calls OmmServerBaseImpl::uninitialize(false,false).
There the mutex "_userLock" is locked and thereafter the method wait() for the dispatch thread is called. This blocks the calling thread until the dispatch thread has terminated.
However, the dispatch thread is blocked in OmmBaseImpl::rsslReactorDispatchLoop(), because it wants to obtain the lock on mutex "_userLock". However, this mutex is currently locked by the thread that calls the destructor of OmmConsumer. Therefore the dispatch thread is blocked and cannot terminate.
So this is a perfect deadlock situation.
Regards, Michael Zuck

**Comment by IProkopenkoRef** [comment]
Hi Michael
Could you provide code snippet for the further investigation so issue could be reproduced? Also please provide more specifics:
a. Linux platform
b. compiler
c. EMA configuration, especially which dispatching model is used API/application

**Comment by vlevendel** [comment]
@miczuc Any feedback?

**Comment by miczuc** [comment]
This problem has nothing to do with the platform or compiler. It depends only on the flow of execution in two threads. I explained in detail, how the deadlock situation occurs. Everyone, who looks at the code, will understand that. However, I noticed the problem on CentOS 7, gcc 4.8. I noticed, that termininating a simple consumer like "Cons100" with Ctrl-C (console interrupt key) sometimes had no effect on the first time, but sometimes had. Then I looked with pstack (= gstack) into the consumer program after pressing Ctrl-C without effect and observed, that there are all threads waiting / blocked. After examining the call stacks of the threads, it became obvious, that this was a deadlock as explained.
